src/lib/locationState/syncers/next-pages/next-pages-syncer.ts

```diff
@@ -0,0 +1,28 @@
+import { Syncer } from "@/lib/locationState/syncers/navigation/types";
+import { NextRouter } from "next/router";
+
+export class NextPagesSyncer implements Syncer {
+  constructor(private readonly router: NextRouter) {}
+
+  key(): string | undefined {
+    return globalThis.history.state.key;
+  }
+
+  sync({
+    listener,
+    signal,
+  }: {
+    listener: (key: string) => void;
+    signal: AbortSignal;
+  }): void {
+    const handler = () => {
+      // Since an Entry always exists at the time of `routeChangeStart`, it is non-null.
+      listener(this.key()!);
+    };
+    this.router.events.on("routeChangeComplete", handler);
```

Comment:
recoil-sync-nextと同じようなやり方できないかと思ったんですが、今の設計だと完全なkeyがそろったタイミングで通知する必要があるので、`history.state`が変化するのを待つ必要があって`routeChangeComplete`にしています。
https://github.com/recruit-tech/recoil-sync-next/blob/main/src/history/useSyncHistory.ts#L90C1-L92C41
やりよう考えてみたんですがなかなか浮かばず停滞してます...

Reply:
memo: ブラウザバック判定して違う時は`store.reset()`+complete時に`store.registerKey(key)`試してみる

Reply:
でも`routeChangeComplete`だと余計なレンダリングが挟まっちゃうと思うんですよね
特にDynamic Routesの場合に顕著になるはず (なので`/ssr/[id]`と`ssg/[id]`をサンプルに入れたい)

もし`ssr/1`で状態`foo`が`"hoge"`の場合に`ssr/2`へナビゲートする場合、

- `routeChangeStart`
- レンダリング (`id: 2`, `foo: "hoge"`)
- `routeChangeComplete`
- 再レンダリング (`id: 2, `foo: ""`)

という挙動になるんじゃないかな？ (サンプルを追加したら確認してみて)
目視できなくてもDevToolsのPerformanceタブでレコードすると`id: 2`, `foo: "hoge"`のフレームがいるんじゃないかと思う
なので`routeChangeComplete`では手遅れだと思うんだよね

Reply:
recoil-sync-nextで参考にすべきはここだったりする
https://github.com/recruit-tech/recoil-sync-next/blob/d893a1e52ed0a67a330dae45a5715ba5632c43a9/src/url/useSyncURLNext.ts#L25

この`handler()`が`Syncer`への通知に相当する
つまりレンダリングのタイミングで通知してるわけ
正確にはrecoil-sync-urlは`useEffect()`で通知してるんだけど、これはrecoilが非同期に対応しているからで、今回のlocation-stateは同期だからその必要もない、と思われ

つまり

- `routeChangeStart`
- レンダリング → ここで`store.load()` → そのままレンダリング (`id: 2`, `foo: ""`)
- `routeChangeComplete`

としたい

Reply:
recoil-sync-nextは`isReady`にも対処してるけどloction-stateは`useSyncExternalStore()`のおかげでそれを考慮する必要はないはず
結構シンプルに実装できるんじゃないかな

Reply:
サンプル追加してみて余計なレンダリングが挟まってしまう点は理解できました。
修正方針ですが、ちょっと僕の理解と解像度がだいぶ低いので相談させてください🙇‍♂️
今の実装から`routeChangeStart`に変更しただけだとkeyの更新がされてないため、待つ必要があると思ってます。
```ts
    const handler = () => {
      // Since an Entry always exists at the time of `routeChangeStart`, it is non-null.
      listener(this.key()!); // <- ここが古い
    };
    this.router.events.on("routeChangeStart", handler);

    signal.addEventListener("abort", () => {
      this.router.events.off("routeChangeStart", handler);
    });
```
Syncerの中はuseEffect使えないので、↑の`handler`の中でlistenerへの通知を遅らせるイメージでいますが合ってますか？
またここで通知遅らせるにはどうやって値の変化を監視すべきか悩んでます。
popStateでは拾いきれなかったので`queueMicrotask`とか`requestAnimationFrame`で再帰で`history.state.key`が変わるのを待つとかかなと思ったんですが、こういちさん的には全然違うイメージを持ってるような気がしており。。。

Reply:
そうですね、もっと簡単にできると思ってます
まずは`Syncer`ですが、Pages Routerの`Router.events`は直接は使えないので、`Syncer`は`listener`を管理するだけにして、それらに通知するメソッドを持たせます
```typescript
export class NextPagesSyncer implements Syncer {
  private readonly listeners = new Set<(key: string) => void>();
  ...
  sync({
    listener,
    signal,
  }: {
    listener: (key: string) => void;
    signal: AbortSignal;
  }): void {
    this.listeners.add(listener);
    signal?.addEventListener("abort", () => {
      this.listeners.delete(listener);
    });
  }

  notify() {
    const currentKey = this.key()!;
    this.listeners.forEach((listener) => listener(currentKey));
  }
}
```
(`key`が前と同じだったら～とかのチェックは省略)

そして実際に通知をする、つまり`notify()`を呼び出すのは先のコメントに書いたようにレンダリング時になります
そこでAppコンポーネントから呼び出されるカスタムhookを作ります
このhookは`Syncer`と連携する必要があるので、戻り値で`Syncer`を返します
ナビゲーション時は必ずAppコンポーネントが実行されるのでこのhookも実行されるというのがポイントです
```typescript
export function useNextPagesSyncer() {
  const needNotify = React.useRef(false);
  const [syncer] = React.useState(() => new NextPagesSyncer());
  if (needNotify.current) {
    syncer.notify();
    needNotify.current = false;
  }

  React.useEffect(() => {
    const routeChangeHandler = () => {
      needNotify.current = true;
    };
    Router.events.on("routeChangeStart", routeChangeHandler);
    return () => Router.events.off("routeChangeStart", routeChangeHandler);
  }, []);

  return syncer;
}
```
`"routeChangeStart"`後のAppコンポーネントのレンダリング時には`history.state.key`もURLも正しく設定されているので、このタイミングで同期的に状態を復元すればその後のPageコンポーネントはいきなり正しい状態でレンダリングできます (SessionStorageが同期でありがたいと思ったのは初めてかも)
本来ハイドレーション時は復元する前の状態でレンダリングしないとSSR/SSGと一致しないので問題になるのですが、`useSyncExternalStore()`はハイドレーションでは`getServerSnapshot()`が返す値 (つまりデフォルト値) でレンダリングしてくれるのでそれを考慮する必要もありません

ということで組み込んでみてください！！

Reply:
8aab74e57d7a0ee9e36b1688f8ebaa6fe1b5c9c6
組み込んでみました！
動かしてみた感じもおっしゃる通りに動作して見えます。
`_app`の再レンダリングのタイミングが`"routeChangeStart"`後はgSSPのjson fetch後になるからrefで書き換えて次のレンダリングを待つってことですよね？そこらへんも分かってなかったので、ようやく合点がいきました。
>  useSyncExternalStore()はハイドレーションではgetServerSnapshot()が返す値 (つまりデフォルト値) でレンダリングしてくれるので

これもなるほどです、全然わかってなかった...
だいぶ理解進んできた気がします、ありがとうございます！🙇‍♂️
（これをさらっと書いてしまうのすごすぎる）